HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc

```diff
@@ -0,0 +1,440 @@
+    if (reco::deltaR2(*r1, *r2) < maxDeltaR2)
+      return true;
+  }
+  return false;
+}
+
+double HLTMuonL3orL2PreFilter::pTeffTransformation(const reco::RecoChargedCandidateRef& candidate,
+                                                   const reco::TrackRef& track,
+                                                   const cutParameters& cutPars) const {
+  auto const pt = candidate->pt();
+  auto const err0 = track->error(0);
+  auto const abspar0 = std::abs(track->parameter(0));
+  // if abspar0 > 0, convert 50% efficiency threshold to 90% efficiency threshold
+  auto const ptLx = abspar0 > 0 ? pt + cutPars.nsigma_Pt * err0 / abspar0 * pt : pt;
+  return ptLx;
+}
+
+bool HLTMuonL3orL2PreFilter::applySelection(const reco::RecoChargedCandidateRef& candidate,
+                                            const reco::BeamSpot& beamSpot,
+                                            const cutParameters& cutPars) const {
+  // eta cut
+  if (std::abs(candidate->eta()) > cutPars.max_Eta)
+    return false;
+
+  reco::TrackRef const& tk = candidate->track();
+  LogDebug("HLTMuonL3orL2PreFilter") << " Muon in loop, q*pt= " << tk->charge() * tk->pt() << " ("
+                                     << candidate->charge() * candidate->pt() << ") "
+                                     << ", eta= " << tk->eta() << " (" << candidate->eta() << ") "
+                                     << ", hits= " << tk->numberOfValidHits() << ", d0= " << tk->d0()
+                                     << ", dz= " << tk->dz();
+
+  // normalizedChi2 cut
+  if (tk->normalizedChi2() > cutPars.max_NormalizedChi2)
+    return false;
+
+  // number of eta bins for cut on number of stations
+  const auto nAbsetaBins = cutPars.absetaBins.size();
+
+  // cut on number of stations
+  bool failNstations(false), failNhits(false), failNchambers(false);
+  for (unsigned int i = 0; i < nAbsetaBins; ++i) {
+    if (std::abs(candidate->eta()) < cutPars.absetaBins[i]) {
+      if (candidate->track()->hitPattern().muonStationsWithAnyHits() < cutPars.minNstations[i]) {
+        failNstations = true;
+      }
+      if (candidate->track()->numberOfValidHits() < cutPars.min_Nhits[i]) {
+        failNhits = true;
+      }
+      if ((candidate->track()->hitPattern().dtStationsWithAnyHits() +
+               candidate->track()->hitPattern().cscStationsWithAnyHits() <
+           cutPars.minNchambers[i])) {
+        failNchambers = true;
+      }
+      break;
+    }
+  }
+  if (failNstations || failNhits || failNchambers)
+    return false;
+
+  // dz cut (impact parameter of the muon in the z direction computed with respect to the beamspot position)
+  auto const vz = candidate->vz() - beamSpot.z0();
+  auto const vx = candidate->vx() - beamSpot.x0();
+  auto const vy = candidate->vy() - beamSpot.y0();
+  if (std::abs(vz - (vx * candidate->px() + vy * candidate->py()) / (candidate->pt() * candidate->pt()) *
+                        candidate->pz()) > cutPars.max_Dz)
```

Comment:
```suggestion
  if (candidate->pt() <= 0 or (std::abs(vz - (vx * candidate->px() + vy * candidate->py()) / (candidate->pt() * candidate->pt()) *
                        candidate->pz()) > cutPars.max_Dz))
```

Reply:
After a 2nd look, I think it would be sufficient to add a condition like the one here to avoid the pT==0 case. If you prefer to keep the explicit check on "pT == 0" as in the previous implementation, that's also okay (but it should probably become a LogWarning instead of a LogDebug).

Reply:
Yes, I think the same. I didn't see this comment